### PR TITLE
tsdb: fix races around head chunks

### PR DESCRIPTION
**Pull request by roidelapluie**
Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

**Comment by roidelapluie** [comment]
Maybe fix #6977

**Comment by roidelapluie** [comment]
this is really a draft and something that comes into my head like, it could have caused the panic we have seen.

**Comment by roidelapluie** [comment]
```
==================
WARNING: DATA RACE
Write at 0x00c000122028 by goroutine 26:
  runtime.slicecopy()
      /home/roidelapluie/godist/go/src/runtime/slice.go:197 +0x0
  github.com/prometheus/prometheus/tsdb.(*memSeries).truncateChunksBefore()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1849 +0x1bd
  github.com/prometheus/prometheus/tsdb.(*stripeSeries).gc()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1643 +0xd5
  github.com/prometheus/prometheus/tsdb.(*Head).gc()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1150 +0x9e
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace.func1()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1630 +0x58

Previous read at 0x00c000122028 by goroutine 9:
  github.com/prometheus/prometheus/tsdb.(*memSeries).minTime()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1769 +0x6a2
  github.com/prometheus/prometheus/tsdb.(*headIndexReader).Postings()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1427 +0x4a9
  github.com/prometheus/prometheus/tsdb.PostingsForMatchers()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:395 +0xf50
  github.com/prometheus/prometheus/tsdb.lookupChunkSeries()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:711 +0xad
  github.com/prometheus/prometheus/tsdb.LookupChunkSeries()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:698 +0xdc
  github.com/prometheus/prometheus/tsdb.(*blockQuerier).Select()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:199 +0x40
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1633 +0x512
  github.com/prometheus/prometheus/tsdb.TestHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1602 +0x41
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb

Goroutine 26 (running) created at:
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1628 +0x495
  github.com/prometheus/prometheus/tsdb.TestHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1602 +0x41
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb

Goroutine 9 (running) created at:
  testing.(*T).Run()
      /home/roidelapluie/godist/go/src/testing/testing.go:1044 +0x660
  testing.runTests.func1()
      /home/roidelapluie/godist/go/src/testing/testing.go:1286 +0xa6
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb
  testing.runTests()
      /home/roidelapluie/godist/go/src/testing/testing.go:1284 +0x527
  testing.(*M).Run()
      /home/roidelapluie/godist/go/src/testing/testing.go:1201 +0x2ff
  main.main()
      _testmain.go:272 +0x223
==================
==================
WARNING: DATA RACE
Write at 0x00c00072e028 by goroutine 26:
  github.com/prometheus/prometheus/tsdb.(*memSeries).truncateChunksBefore()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1849 +0x1ce
  github.com/prometheus/prometheus/tsdb.(*stripeSeries).gc()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1643 +0xd5
  github.com/prometheus/prometheus/tsdb.(*Head).gc()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1150 +0x9e
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace.func1()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1630 +0x58

Previous read at 0x00c00072e028 by goroutine 9:
  github.com/prometheus/prometheus/tsdb.(*memSeries).minTime()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1766 +0x48b
  github.com/prometheus/prometheus/tsdb.(*headIndexReader).Postings()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1427 +0x4a9
  github.com/prometheus/prometheus/tsdb.PostingsForMatchers()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:395 +0xf50
  github.com/prometheus/prometheus/tsdb.lookupChunkSeries()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:711 +0xad
  github.com/prometheus/prometheus/tsdb.LookupChunkSeries()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:698 +0xdc
  github.com/prometheus/prometheus/tsdb.(*blockQuerier).Select()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:199 +0x40
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1633 +0x512
  github.com/prometheus/prometheus/tsdb.TestHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1602 +0x41
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb

Goroutine 26 (running) created at:
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1628 +0x495
  github.com/prometheus/prometheus/tsdb.TestHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1602 +0x41
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb

Goroutine 9 (running) created at:
  testing.(*T).Run()
      /home/roidelapluie/godist/go/src/testing/testing.go:1044 +0x660
  testing.runTests.func1()
      /home/roidelapluie/godist/go/src/testing/testing.go:1286 +0xa6
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb
  testing.runTests()
      /home/roidelapluie/godist/go/src/testing/testing.go:1284 +0x527
  testing.(*M).Run()
      /home/roidelapluie/godist/go/src/testing/testing.go:1201 +0x2ff
  main.main()
      _testmain.go:272 +0x223
==================
==================
WARNING: DATA RACE
Write at 0x00c00072e040 by goroutine 26:
  github.com/prometheus/prometheus/tsdb.(*memSeries).truncateChunksBefore()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1854 +0x36b
  github.com/prometheus/prometheus/tsdb.(*stripeSeries).gc()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1643 +0xd5
  github.com/prometheus/prometheus/tsdb.(*Head).gc()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1150 +0x9e
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace.func1()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1630 +0x58

Previous read at 0x00c00072e040 by goroutine 9:
  github.com/prometheus/prometheus/tsdb.(*memSeries).head()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:2015 +0x5fa
  github.com/prometheus/prometheus/tsdb.(*memSeries).maxTime()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1773 +0x5e5
  github.com/prometheus/prometheus/tsdb.(*headIndexReader).Postings()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head.go:1427 +0x4a9
  github.com/prometheus/prometheus/tsdb.PostingsForMatchers()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:395 +0xf50
  github.com/prometheus/prometheus/tsdb.lookupChunkSeries()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:711 +0xad
  github.com/prometheus/prometheus/tsdb.LookupChunkSeries()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:698 +0xdc
  github.com/prometheus/prometheus/tsdb.(*blockQuerier).Select()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/querier.go:199 +0x40
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1633 +0x512
  github.com/prometheus/prometheus/tsdb.TestHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1602 +0x41
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb

Goroutine 26 (running) created at:
  github.com/prometheus/prometheus/tsdb.testHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1628 +0x495
  github.com/prometheus/prometheus/tsdb.TestHeadSeriesChunkRace()
      /home/roidelapluie/go/src/github.com/prometheus/prometheus/tsdb/head_test.go:1602 +0x41
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb

Goroutine 9 (running) created at:
  testing.(*T).Run()
      /home/roidelapluie/godist/go/src/testing/testing.go:1044 +0x660
  testing.runTests.func1()
      /home/roidelapluie/godist/go/src/testing/testing.go:1286 +0xa6
  testing.tRunner()
      /home/roidelapluie/godist/go/src/testing/testing.go:993 +0x1eb
  testing.runTests()
      /home/roidelapluie/godist/go/src/testing/testing.go:1284 +0x527
  testing.(*M).Run()
      /home/roidelapluie/godist/go/src/testing/testing.go:1201 +0x2ff
  main.main()
      _testmain.go:272 +0x223
==================
--- FAIL: TestHeadSeriesChunkRace (29.20s)
    testing.go:907: race detected during execution of test
FAIL
exit status 1
FAIL	github.com/prometheus/prometheus/tsdb	29.240s
```

**Comment by roidelapluie** [comment]
Just want to see the overhead of locking. Will shut it down in the morning.

File: tsdb/head.go [hunk 1]
```diff
@@ -1738,6 +1738,7 @@ type memSeries struct {
 	ref          uint64
 	lset         labels.Labels
 	chunks       []*memChunk
+	chunksLock   sync.RWMutex
```
Comment (codesome) [inline tsdb/head.go]: The `memSeries` itself is a `Mutex` which is held before calling `truncateChunksBefore` and `append`. Can we change it to `RWMutex` and use that instead?

Comment (roidelapluie) [reply tsdb/head.go]: sure thing, done :)

**Comment by roidelapluie** [comment]
While this *technically* fixes the race at a golang level, what does it truly mean for #6676 when we have garbage collection while we read the postings?

**Comment by roidelapluie** [comment]
goroutine dump

https://gist.github.com/roidelapluie/95a71d16b7c10be8bd8eee01694e1fe2

**Comment by roidelapluie** [comment]
Let's get a rc.1 out with this and see.

**Pull request merged**